### PR TITLE
Add documentation for cancellation of request processing

### PR DESCRIPTION
**Pull request by squahtx**
[Rendered](https://github.com/matrix-org/synapse/blob/squah/add_cancellation_docs/docs/development/synapse_architecture/cancellation.md)

Please call out anything that is unclear or confusing. If you have questions, it's a sign that the answers are missing from the docs!

### CODE REVIEW

File: docs/cancellation.md [hunk 1]
```diff
@@ -0,0 +1,387 @@
+# Cancellation
+Sometimes, requests take a long time to service and clients disconnect
+before Synapse produces a response. To avoid wasting resources, Synapse
+can cancel request processing for select endpoints with the `@cancelled`
```
Comment (anoadragon453) [inline docs/cancellation.md:4]:
```suggestion
can cancel request processing for select endpoints with the `@cancellable`
```

**Review by anoadragon453 — APPROVED** [review]
So, does `@cancellable` actually do anything other than serve as a marker? I suppose we don't really have any way of checking?

Otherwise, this mostly made sense to someone who wasn't very familiar with cancellation up until this point. Thanks for writing it up!

File: docs/SUMMARY.md [hunk 2]
```diff
@@ -89,6 +89,7 @@
   - [Database Schemas](development/database_schema.md)
   - [Experimental features](development/experimental_features.md)
   - [Synapse Architecture]()
+    - [Cancellation](cancellation.md)
```
[the remark this answers is not in the document]
Comment (squahtx) [reply docs/SUMMARY.md]: Fixed! I was misled by the existing documentation files 😢

File: docs/cancellation.md (hunk 1 above)
Comment (squahtx) [reply docs/cancellation.md]: Good spot, let's fix that.

I'll write some more words on the decorator.

**Comment by squahtx** [comment]
Thanks for the review! I'm glad it mostly made sense.

> So, does `@cancellable` actually do anything other than serve as a marker? I suppose we don't really have any way of checking?

It's merely a marker. I'm not sure I follow your second question.

**Comment by squahtx** [comment]
> > > So, does @cancellable actually do anything other than serve as a marker? I suppose we don't really have any way of checking?
> 
> > I'm not sure I follow your second question.
> 
> Sorry, I meant that there's not really any code we could add to the `@cancellable` decorator to ensure cancellation is properly handled in the associated function? It's instead a purely manual process?

That's right, it's purely manual checking.
If we were clever, I suppose we could come up with some linter rules which would cover some common cases. There'd be no way to make them exhaustive and cover everything.

File: docs/development/synapse_architecture/cancellation.md [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+    await d
+    return 200, {}
+
+request = defer.ensureDeferred(on_GET())
+```
+
+When a client disconnects early, Synapse checks for the presence of the
+`@cancellable` decorator on `on_GET`. Since `on_GET` is cancellable,
+`Deferred.cancel()` is called on the `Deferred` from
+`defer.ensureDeferred`, `request`. Twisted knows which `Deferred`
```
Comment (clokep) [inline docs/development/synapse_architecture/cancellation.md:53]: I'm finding the second sentence really difficult to flow, what's the ", `request`" hanging at the end?

Comment (squahtx) [reply docs/development/synapse_architecture/cancellation.md:53]: ", ie. `request`." might be clearer?

Comment (clokep) [reply docs/development/synapse_architecture/cancellation.md:53]: Ah, yeah that is clearer. 👍

**Comment by squahtx** [comment]
Thanks for reviewing and proofreading!

**Pull request merged**